### PR TITLE
Non-wrapping mode for better Transformers compatibility

### PR DESCRIPTION
**Pull request by evgri243**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs change / refactoring / dependency upgrade

## Motivation and Context / Related issue

The primary goal of this PR is to introduce a **Non-wrapping mode** (controller-based) to Opacus. This feature allows for the computation of per-sample gradients without wrapping the original `nn.Module` in a `GradSampleModule` subclass. 

This change addresses critical compatibility issues with third-party libraries—most notably HuggingFace `Transformers` and `Accelerate`—which often rely on strict module hierarchies or perform type checks that are disrupted by standard Opacus wrappers.

**Key Technical Changes:**
- **`GradSampleHooks` Controller**: Introduced a new architecture where gradient sampling hooks are managed by a standalone `GradSampleHooks` object instead of being embedded in a `GradSampleModule` wrapper.
- **`PrivacyEngine` Updates**: Modified `make_private` and `make_private_with_epsilon` to support a `wrap_model` parameter (defaults to `True` for backward compatibility). When `wrap_model=False`, Opacus attaches hooks to the original model and returns it along with a `GradSampleHooks` controller.
- **Enhanced Loss Compatibility**: Updated `DPLossFastGradientClipping` to handle custom loss implementations that lack a `reduction` attribute (a common occurrence in the `transformers` library). The implementation now explicitly sets the expected reduction if missing and restores the previous state after the backward pass.
- **HuggingFace Trainer Integration**: Verified the new mode via successful small-scale training runs using the HuggingFace `Trainer` with Fast Gradient Clipping. A new example (`examples/huggingface_trainer.py`) is included to demonstrate this integration.
- **Architectural Cleanup**: Refactored `gsm_base.py` to provide a unified `AbstractGradSampleHooks` base class, improving the robustness of hook attachment, attribute cleanup, and validation logic across all sampling modes.

## How Has This Been Tested (if it applies)

- **New Unit Tests**: 
- **Integration Testing**: Successfully performed training with the HuggingFace `Trainer` using the new non-wrapping mode for both next-token prediction and SFT/DPO.

## Checklist

- [x] The documentation is up-to-date with the changes I made (including a new tutorial and updated READMEs).
- [x] I have read the **CONTRIBUTING** document and completed the CLA.
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by iden-kalemaj** [comment]
That's a good point... let me know how it's looking once you work on it.

**Comment by evgri243** [comment]
Ok. Calling it "as limited as possible" is an overstatement, but I've drastically changed the design and did my best to avoid unnecessary changes

File: opacus/grad_sample/grad_sample_module_fast_gradient_clipping.py [hunk 1]
```diff
@@ -40,31 +40,32 @@
...
+            grad_sample = torch.stack(grad_sample, dim=0).sum(dim=0)
+
         if (
             max_batch_len == 0
         ):  # To handle the case of empty batch that may arise from Poisson sampling
             param._norm_sample = torch.tensor(
                 [], device=grad_sample.device, dtype=grad_sample.dtype
             )
         else:
-            param._norm_sample = torch.zeros(
```
Comment (iden-kalemaj) [inline opacus/grad_sample/grad_sample_module_fast_gradient_clipping.py]: what is the reason for removing this? this was added for low-precision and mixed-precision functionality.

File: opacus/tests/mixed_precision_test.py [hunk 2]
```diff
@@ -73,27 +74,34 @@ def _get_training_components(
 
         # Make the model private with the specified precision
         if grad_sample_mode in ["hooks", "functorch", "ew"]:
-            model, optimizer, dataloader = privacy_engine.make_private(
+            hooks_or_wrapper, optimizer, dataloader = privacy_engine.make_private(
```
Comment (iden-kalemaj) [inline opacus/tests/mixed_precision_test.py]: nit: call it hooks_or_gs_module

File: examples/huggingface_trainer.py [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+HuggingFace Trainer with Differential Privacy support.
+This example demonstrates a robust integration of Opacus with HuggingFace Trainer,
+following the original implementation used in federated research.
```
[the remark this answers is not in the document]
Comment (evgri243) [reply examples/huggingface_trainer.py]: Removed and paraphrased. It slipped...

**Comment by evgri243** [comment]
I  addressing your cutting all unnecessary changes surgically. Should be out first half next week.

**Review by iden-kalemaj — APPROVED** [review]
Thank you again @evgri243 for contributing this heavy lift functionality. I've left a few minor comments, but this looks good to me now. 

Just to check, were you able to run the multi-GPU unit tests in a multi-gpu environment? Github CI does not support the multi-gpu tests for now. If not, I can run them.

**Review by iden-kalemaj — COMMENTED** [review]
hi @evgri243, did you intend to submit this PR_MERGE_PLAN?

**Comment by evgri243** [comment]
Ofc not... a short moment of shame.

Back to changes. I squashed all renaming back and forth and returned to what it used to be + some changes that popped up during test stabilization.

**Comment by iden-kalemaj** [comment]
@evgri243 no worries, can you rebase to main and resolve conflicts after we landed some changes from your team?

let's try to land your PR soon after that

**Comment by evgri243** [comment]
I guess I forgot to push. I did that

**Comment by iden-kalemaj** [comment]
@evgri243 for some reason the system is failing to import the PR internally. I'll check back on this tomorrow morning.

**Comment by iden-kalemaj** [comment]
@evgri243  can you try rebasing and seeing if there are any conflicts, we are unable to import the PR internally which is usually due to merge conflicts.

**Comment by evgri243** [comment]
Rebased and pushed. Let's see.

PS. I am so slow as I am not happy with the resulting diamond inheritance between module and hooks. But so far I fail to think of a better way.

If we get it merged, may I ask you to trigger a build release with all the changes we submitted. This is exactly what was required to make our trainer work. We will rebase it on upstream Opacus and test it thoroughly in the training pipeline before we make further decisions. What do you think?

**Comment by iden-kalemaj** [comment]
@evgri243 that sounds good since we're due for another version release. The current solution looks elegant enough :)

Please see the lint failure and to speed things up please test with all 3 linters [here](https://github.com/meta-pytorch/opacus/blob/main/CONTRIBUTING.md#our-development-process).

**Comment by evgri243** [comment]
Interestingly. It was something old. Fixed it

**Comment by evgri243** [comment]
See massive failures. Attending to them today.

**Comment by evgri243** [comment]
CI/CD seems to be finally green. We rebased our code on top of this branch and previously merged changes -- seems working, but we will continue thorough testing.

**Comment by iden-kalemaj** [comment]
@evgri243 i did not get to check the initial failures, can you let me know why the tests were failing? Just want to make sure that we are not duplicating any of the code in #805

**Comment by evgri243** [comment]
The errors were just because I removed all the changes with arithmetics. This PR had them originally and then removed in favor of an independent PR. Seeing that, rebase silently undid them as well
